src/movies.js

```diff
@@ -1,16 +1,107 @@
 // Iteration 1: All directors? - Get the array of all directors.
+
+function getAllDirectors(moviesArr) {
+    let directorsArr = moviesArr.map((el, ind) => 
+    {return moviesArr[ind].director})
+    return directorsArr
+}
+
 // _Bonus_: It seems some of the directors had directed multiple movies so they will pop up multiple times in the array of directors. How could you "clean" a bit this array and make it unified (without duplicates)?
 
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
 
+function howManyMovies(moviesArr) {
+    let spielbergDramas = moviesArr.filter ((el, ind) => {
+        return moviesArr[ind].director == "Steven Spielberg" && moviesArr[ind].genre.includes("Drama")
+        }
+    )
+    return spielbergDramas.length
+}
+
 // Iteration 3: All rates average - Get the average of all rates with 2 decimals
 
+function ratesAverage(moviesArr) {
+  let newMoviesArr = []
+  for (i = 0; i < moviesArr.length; i++) {
+    newMoviesArr.push(moviesArr[i])
+  }
+    if(newMoviesArr.length == 0) {
+        return 0
+    } else {
+      for (let i = 0; i < newMoviesArr.length; i++) {
+/*
+This next line should fix jasmine issue number 17: "Get the average 
+rate - ratesAverage Return average even if one of the movies does 
+not have rate!". But it doesn't, and I don't know why
+*/
+        if (typeof (newMoviesArr[i].rate) !== "number") {
+          newMoviesArr.splice(i, 1)
+        }
+      }
+    let moviesTotalScore = newMoviesArr.reduce((acc, val) => {
+        return acc + val.rate;
+    }, 0);
+    return Math.round((moviesTotalScore / newMoviesArr.length)*100)/100
```

Comment:
Entiendo la idea que has tenido aquí para limpiar al array de los elemento que el rate no tenga un valor numérico, pero tienes un problema, si recordamos la propiedad `splice` lo que hace es modificar el array original, en este caso eliminando un elemnto, por tanto si tenemos un array de tres elementos `['pepe', 'luis', 'lolo']`, y queremos eliminar cuando el elemento valga `luis` que es en la segunda posición. Cuando estamos recorriendo el array con un for utilizando un index, cuando eliminemos el elemento `luis` el array tendrá de longitud `2`, por tanto al pasar al siguiente paso nos saltaremos el tercer elemento.
```
const arr = ['pepe', 'luis', 'lolo'];

for (let i = 0; i < arr.length; i++) {
  if (arr[i] === 'luis') {
    arr.slice(i, 1);  
  }
}
```
Ahora vamos a ir parte por parte:
i = 0
```
for (let i = 0; i < arr.length; i++) { // --> arr.length = 2
  if (arr[0] === 'luis') { // 'pepe' es diferente a 'luis', por tanto no se ejecuta
    arr.slice(i, 1);  
  } // --> arr.lengh = 2
}
```
i = 1
```
for (let i = 0; i < arr.length; i++) { // --> arr.length = 2
  if (arr[0] === 'luis') { // 'luis' es igual a 'luis', por tanto se ejecuta eliminando ese elemento
    arr.slice(i, 1);  
  } // --> arr.length = 1
}
```
i = 2
```
for (let i = 0; i < arr.length; i++) { // --> arr.length = 1, entonces la condicion 2 < 1 no se cumple, por tanto el for no continua y no verifica el siguiente valor
  if (arr[0] === 'luis') { 
    arr.slice(i, 1);  
  }
}
```
Por eso no se recomienda utilizar un `splice` dentro de un for que recorre ese mismo array, para eso podemos utilizar `filter`. Igualmente esto lo podríamos solucionar añadiento un i-- despues de la ejecución del `splice`:

```
const arr = ['pepe', 'luis', 'lolo'];

for (let i = 0; i < arr.length; i++) {
  if (arr[i] === 'luis') {
    arr.slice(i, 1);  
    i--; // <---
  }
}
```